### PR TITLE
Migrate gRPC VerificationService

### PR DESCRIPTION
**Pull request by mustard-mh**
## Description

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at 8ba3e63</samp>

No summary available (Limit exceeded: required to process 106320 tokens, but only 50000 are allowed per call)

</details>

## Related Issue(s)

Fixes EXP-846

## How to test

Test 1

Create a fresh GitHub account and sign in (use admin to unblock).
You should need to do phone number verification before you can start your first workspace.
Once the number is verified you should be able to work as usual.

Test 2

Use a GitHub account older than 30 days.
You should be able to start a workspace without verification.

Test 1 ++

delete the lastVerificationTimestamp in your user's additional Data.
if the account had started a workspace before, you should not need to verify
if the account doesn't have any workspaces, you should be asked to verify

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - hw-papi-auth</li>
	<li><b>🔗 URL</b> - <a href="https://hw-papi-auth.preview.gitpod-dev.com/workspaces" target="_blank">hw-papi-auth.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - hw-papi-auth-gha.20835</li>
	<li><b>🗒️ Logs</b> - <a href="https://console.cloud.google.com/logs/query;query=jsonPayload.kubernetes.host%3D%22preview-hw-papi-auth%22%0A%0A--%20Filter%20on%20service:%0A--%20jsonPayload.serviceContext.service%3D%22ws-manager-mk2%22%0A;duration=P1D?project=gitpod-core-dev" target="_blank">GCP Logs Explorer</a></li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [x] /werft preemptible
      Saves cost. Untick this only if you're really sure you need a non-preemtible machine.
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

File: components/server/src/api/verification-service-api.ts [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+import { PaginationToken, generatePaginationToken, parsePaginationToken } from "./pagination";
+import { parseSorting } from "./sorting";
+import { PaginationResponse } from "@gitpod/public-api/lib/gitpod/v1/pagination_pb";
+import { PublicAPIConverter } from "@gitpod/gitpod-protocol/lib/public-api-converter";
+
+@injectable()
+export class VerificationServiceAPI implements ServiceImpl<typeof VerificationServiceInterface> {
+    @inject(VerificationService) private readonly verificationService: VerificationService;
+    @inject(UserService) private readonly userService: UserService;
+    @inject(UserDB) private readonly userDB: UserDB;
```
Comment (akosyakov) [inline components/server/src/api/verification-service-api.ts:43]: I hope we don't need to leak DB objects outside of app services anymore.

File: components/server/src/api/verification-service-api.ts [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+        const { verified } = await this.verificationService.verifyVerificationToken(
+            phoneNumber,
+            req.token,
+            req.verificationId,
+        );
+        if (verified) {
+            this.verificationService.markVerified(user);
+            user.verificationPhoneNumber = phoneNumber;
+            await this.userDB.updateUserPartial(user);
```
Comment (akosyakov) [inline components/server/src/api/verification-service-api.ts:102]: should not we use userService instead?

File: components/server/src/auth/verification-service.ts [hunk 3]
```diff
@@ -4,24 +4,30 @@
...
+import { Authorizer } from "../authorization/authorizer";
+import { BlockedRepositoryDB } from "@gitpod/gitpod-db/lib/blocked-repository-db";
 
 @injectable()
 export class VerificationService {
     @inject(Config) protected config: Config;
     @inject(WorkspaceDB) protected workspaceDB: WorkspaceDB;
     @inject(UserDB) protected userDB: UserDB;
     @inject(TeamDB) protected teamDB: TeamDB;
+    @inject(Authorizer) private readonly auth: Authorizer;
```
Comment (akosyakov) [inline components/server/src/auth/verification-service.ts:28]: @svenefftinge @geropl is it a good place?

File: components/server/src/auth/verification-service.ts [hunk 4]
```diff
@@ -156,4 +162,45 @@ export class VerificationService {
             channel: verification_check.channel,
         };
     }
+
+    public async adminGetBlockedRepositories(
```
Comment (akosyakov) [inline components/server/src/auth/verification-service.ts:166]: all new methods just a move or new? We need DB tests if new.

File: components/dashboard/src/service/public-api.ts [hunk 5]
```diff
@@ -54,6 +56,7 @@ export const workspacesService = createPromiseClient(WorkspaceV1Service, transpo
 export const oidcService = createPromiseClient(OIDCService, transport);
 
 export const workspaceClient = createServiceClient(WorkspaceService, new JsonRpcWorkspaceClient());
+export const verificationClient = createServiceClient(VerificationService, new JsonRpcVerificationClient());
```
Comment (akosyakov) [inline components/dashboard/src/service/public-api.ts:59]: please rebase there should be `verification` suffix 🙏

File: components/public-api/gitpod/v1/verification.proto [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+  // SendPhoneNumberVerificationToken sends a verification token to the
+  // specified phone number.
+  rpc SendPhoneNumberVerificationToken(SendPhoneNumberVerificationTokenRequest) returns (SendPhoneNumberVerificationTokenResponse) {}
+
+  // VerifyPhoneNumberVerificationToken verifies the specified verification
+  // token.
+  rpc VerifyPhoneNumberVerificationToken(VerifyPhoneNumberVerificationTokenRequest) returns (VerifyPhoneNumberVerificationTokenResponse) {}
+
+  // ListBlockedRepositories lists blocked repositories.
+  rpc ListBlockedRepositories(ListBlockedRepositoriesRequest) returns (ListBlockedRepositoriesResponse) {}
```
Comment (akosyakov) [inline components/public-api/gitpod/v1/verification.proto:21]: @mustard-mh I chatted with @AlexTugarev and @geropl about ti, management of blocked repositories and email domains should go to a separate InstallationService. I updated the table.

Comment (mustard-mh) [reply components/public-api/gitpod/v1/verification.proto:21]: @jeanp413  new service name will be `InstallationService` ☝️

File: components/server/src/auth/verification-service.ts [hunk 7]
```diff
@@ -54,35 +58,22 @@ export class VerificationService {
         return isPhoneVerificationEnabled;
     }
 
-    public markVerified(user: User): User {
-        user.lastVerificationTime = new Date().toISOString();
-        return user;
-    }
-
     public async verifyOrgMembers(organizationId: string): Promise<void> {
```
Comment (akosyakov) [inline components/server/src/auth/verification-service.ts:61]: is it fine that we are changing the logic here, are there DB tests for this service? If we are just moving functions without changing semantic is fine with me without adding tests, but if we start changing code we should rather start adding tests.

File: components/server/src/api/verification-service-api.ts [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        // Check if verify via call is enabled
+        const phoneVerificationByCall = await getExperimentsClientForBackend().getValueAsync(
+            "phoneVerificationByCall",
+            false,
+            {
+                user,
+            },
+        );
+        const channel = phoneVerificationByCall ? "call" : "sms";
+        const { verificationId } = await this.verificationService.sendVerificationToken(
```
Comment (akosyakov) [inline components/server/src/api/verification-service-api.ts:49]: we are missing analytics enent here which is present in JSON-RPC

I think much of code from JSON-RPC should have been moved to VerificationService and not copied in API layer, here we shoudl not have logic but rather only translations

File: components/server/src/api/verification-service-api.ts [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+        if (!req.phoneNumber) {
+            throw new ApplicationError(ErrorCodes.BAD_REQUEST, "phoneNumber is required");
+        }
+        if (!req.verificationId || !uuidValidate(req.verificationId)) {
+            throw new ApplicationError(ErrorCodes.BAD_REQUEST, "verificationId is required");
+        }
+        if (!req.token) {
+            throw new ApplicationError(ErrorCodes.BAD_REQUEST, "token is required");
+        }
+        const phoneNumber = formatPhoneNumber(req.phoneNumber);
```
Comment (akosyakov) [inline components/server/src/api/verification-service-api.ts:71]: it should be part of verificaiton services, not copied pasted

**Review by svenefftinge — APPROVED** [review]
Code changes look good. I didn't test, though

File: components/server/src/auth/verification-service.ts (hunk 7 above)
Comment (jeanp413) [reply components/server/src/auth/verification-service.ts:61]: It's just moving functions

**Pull request merged**